File: animation/src/main/java/me/wyne/wutils/animation/BlockingAnimationStep.java
package me.wyne.wutils.animation;

import org.bukkit.Bukkit;
import org.javatuples.Pair;
import org.jetbrains.annotations.NotNull;

public class BlockingAnimationStep extends BaseAnimationStep {

    public BlockingAnimationStep(AnimationRunnable runnable, long delay, long period, long duration) {
        super(runnable, delay, period, duration);
    }

    public BlockingAnimationStep(AnimationRunnable runnable, long delay) {
        super(runnable, delay);
    }

    public BlockingAnimationStep(AnimationRunnable runnable) {
        super(runnable);
    }

    @Override
    protected void runOnce(@NotNull Animation animation) {
        var task = Bukkit.getScheduler().runTaskLater(
                animation.getPlugin(),
                () -> {
                    getRunnable().run(getDelay(), getPeriod(), getDuration());
                    close();
                    startNext(animation);
        }, getDelay());
        animation.setCurrentTask(new Pair<>(this, task));
    }

    @Override
    protected void runRepeating(Animation animation) {
        var task = Bukkit.getScheduler().runTaskTimer(
                animation.getPlugin(),
                () -> {
                    if (getDuration() > 0 && ticksElapsed >= getDuration()) {
                        close();
                        var currentTask = animation.getCurrentTask();
                        if (currentTask != null)
                            currentTask.getValue1().cancel();
                        startNext(animation);
                        return;
                    }
                    getRunnable().run(getDelay(), getPeriod(), getDuration());
                    ticksElapsed += getPeriod();
        }, getDelay(), getPeriod());
        animation.setCurrentTask(new Pair<>(this, task));
    }

}

File: animation/src/main/kotlin/me/wyne/wutils/animation/BlockingAnimationStep.kt
package me.wyne.wutils.animation

import org.bukkit.Bukkit

class BlockingAnimationStep(
    runnable: AnimationRunnable,
    delay: Long = 0,
    period: Long = 0,
    duration: Long = 0
) : BaseAnimationStep(runnable, delay, period, duration) {
    override fun runOnce(animation: Animation) {
        val task = Bukkit.getScheduler().runTaskLater(
            animation.plugin,
            Runnable {
                runnable.run(delay, period, duration)
                close()
                animation.pollStep()?.run(animation)
            },
            delay
        )
        animation.currentTask = this to task
    }

    override fun runRepeating(animation: Animation) {
        val task = Bukkit.getScheduler().runTaskTimer(
            animation.plugin,
            Runnable {
                if (duration > 0 && ticksElapsed >= duration) {
                    close()
                    animation.currentTask?.second?.cancel()
                    animation.pollStep()?.run(animation)
                    return@Runnable
                }
                runnable.run(delay, period, duration)
                ticksElapsed += period
            },
            delay,
            period
        )
        animation.currentTask = this to task
    }
}
